repos/system_upgrade/el7toel8/actors/peseventsscanner/libraries/library.py

```diff
@@ -324,46 +350,54 @@ def add_output_pkgs_to_transaction_conf(transaction_configuration, events):
-    pes_based_pkgs_not_to_be_removed = []
-
-    for pes_based_pkg_to_install in to_install.keys():
-        if pes_based_pkg_to_install in transaction_configuration.to_remove:
-            to_install.pop(pes_based_pkg_to_install)
-            pes_based_pkgs_not_to_be_installed.append(pes_based_pkg_to_install)
-    for pes_based_pkg_to_remove in to_remove.keys():
-        if pes_based_pkg_to_remove in transaction_configuration.to_install + transaction_configuration.to_keep:
-            to_remove.pop(pes_based_pkg_to_remove)
-            pes_based_pkgs_not_to_be_removed.append(pes_based_pkg_to_remove)
-
-    if pes_based_pkgs_not_to_be_installed:
+    pkgs_not_to_be_kept = []
+    pkgs_not_to_be_installed = []
+    pkgs_not_to_be_removed = []
+
+    for pkg_to_keep in tasks['to_keep'].keys():
+        if pkg_to_keep in transaction_configuration.to_remove:
+            pkgs_not_to_be_kept.append(pkg_to_keep)
+    for pkg_to_install in tasks['to_install'].keys():
+        if pkg_to_install in transaction_configuration.to_remove:
+            pkgs_not_to_be_installed.append(pkg_to_install)
+    for pkg_to_remove in tasks['to_remove'].keys():
+        if pkg_to_remove in transaction_configuration.to_install + transaction_configuration.to_keep:
+            pkgs_not_to_be_removed.append(pkg_to_remove)
+
+    for pkg in pkgs_not_to_be_kept:
+        # Removing a package from the to_keep dict may cause that some repositories won't get enabled
+        tasks['to_keep'].pop(pkg)
+
+    if pkgs_not_to_be_installed:
+        for pkg in pkgs_not_to_be_installed:
+            tasks['to_install'].pop(pkg)
         api.current_logger().debug('The following packages will not be installed because of the'
                                    ' /etc/leapp/transaction/to_remove transaction configuration file:'
-                                   '\n- ' + '\n- '.join(pes_based_pkgs_not_to_be_installed))
-    if pes_based_pkgs_not_to_be_removed:
+                                   '\n- ' + '\n- '.join(pkgs_not_to_be_installed))
+    if pkgs_not_to_be_removed:
+        for pkg in pkgs_not_to_be_removed:
+            tasks['to_remove'].pop(pkg)
         api.current_logger().debug('The following packages will not be removed because of the to_keep and to_install'
                                    ' transaction configuration files in /etc/leapp/transaction/:'
-                                   '\n- ' + '\n- '.join(pes_based_pkgs_not_to_be_removed))
+                                   '\n- ' + '\n- '.join(pkgs_not_to_be_removed))
 
 
-def produce_messages(to_install, to_remove):
-    to_install_pkgs = set(to_install.keys())
-    to_remove_pkgs = set(to_remove.keys())
-    to_enable_repos = set(to_install.values())
+def produce_messages(tasks):
+    to_install_pkgs = list(tasks['to_install'].keys())
+    to_remove_pkgs = list(tasks['to_remove'].keys())
+    to_enable_repos = set(tasks['to_install'].values() + tasks['to_keep'].values())
```

Comment:
What about to change it to list already as well? So no need no need to do that later.
```suggestion
    to_enable_repos = list(set(tasks['to_install'].values() + tasks['to_keep'].values()))
```

Reply:
Done.